File: src/chipschallenge/blockreactions/GreenLockTo.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Inventory.Key;

public class GreenLockTo extends NoSlipReaction {

    private GreenLockTo() {
    }
    private static GreenLockTo mInstance = null;

    public static synchronized GreenLockTo getInstance() {
        if (mInstance == null) {
            mInstance = new GreenLockTo();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        useKey(Key.GREEN);
        standing.destroy();
    }

    public boolean canMove(Block moving, Block standing) {
        return moving.isChip() && hasKey(Key.GREEN);
    }
}

File: ChipsChallenge.Shared/Blockreactions/GreenLockTo.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    using Key = Inventory.Key;

    public class GreenLockTo : NoSlipReaction
    {
        private static GreenLockTo instance;

        private GreenLockTo()
        {
        }

        public static GreenLockTo Instance
        {
            get
            {
                lock (typeof(GreenLockTo))
                {
                    return instance ?? (instance = new GreenLockTo());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
            UseKey(Key.GREEN);
            standing.Destroy();
        }

        public override bool canMove(Block moving, Block standing)
        {
            return moving.Chip && HasKey(Key.GREEN);
        }
    }
}